filebeat/beater/filebeat.go

```diff
@@ -56,46 +55,37 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 
 	var err error
 
-	// Init channels
-	fb.publisherChan = make(chan []*input.FileEvent, 1)
+	// Channel from harvesters to spooler
+	publisherChan := make(chan []*input.FileEvent, 1)
```

Comment:
publisherChan is only required by spooler and publisher. With spooler pushing to publisher can we:
- initialize publisher before spooler
- remove publisherChan and instead give some callback/interface into spooler?

Thinking about init/stop order. How about using `defer X.Stop()` right after creating some async component  `(*Filebeat).Run()` method. Like:

```
X, err := X.New(...)
if err != nil {
    logp.Err(...)
    return err
}
defer X.Stop()
```

The Run method blocks until `fb.done` is closed, which is the case when `(*Filebeat).Stop()` is called. Advantage I see here:
- if some component can not be initialized, all components already started will be closed on shutdown
- makes it easier to maintain stack-like setup/shutdown order due to all Stop methods being close to initialization